jiant/pytorch_transformers_interface/modules.py

```diff
@@ -0,0 +1,300 @@
+import logging as log
+import os
+from typing import Dict
+
+import torch
+import torch.nn as nn
+from allennlp.modules import scalar_mix
+
+import pytorch_transformers
+
+from jiant.preprocess import parse_task_list_arg
+from jiant.utils import utils
+
+
+class PytorchTransformersEmbedderModule(nn.Module):
+    """ Shared code for pytorch_transformers wrappers.
+
+    Subclasses share a good deal of code, but have a number of subtle differences due to different
+    APIs from pytorch_transfromers.
+    """
+
+    def __init__(self, args):
+        super(PytorchTransformersEmbedderModule, self).__init__()
+
+        self.cache_dir = os.getenv(
+            "PYTORCH_PRETRAINED_BERT_CACHE",
+            os.path.join(args.exp_dir, "pytorch_transformers_cache"),
+        )
+        utils.maybe_make_dir(self.cache_dir)
+
+        self.embeddings_mode = args.pytorch_transformers_embedding_mode
+
+        # Integer token indices for special symbols.
+        self._sep_id = None
+        self._cls_id = None
+        self._pad_id = None
+
+        # If set, treat these special tokens as part of input segments other than A/B.
+        self._SEG_ID_CLS = None
+        self._SEG_ID_SEP = None
+
+    def parameter_setup(self, args):
+        # Set trainability of this module.
+        for param in self.model.parameters():
+            param.requires_grad = bool(args.transfer_paradigm == "finetune")
+
+        # Configure scalar mixing, ELMo-style.
+        if self.embeddings_mode == "mix":
+            if args.transfer_paradigm == "frozen":
+                log.warning(
+                    "NOTE: pytorch_transformers_embedding_mode='mix', so scalar "
+                    "mixing weights will be fine-tuned even if BERT "
+                    "model is frozen."
+                )
+            # TODO: if doing multiple target tasks, allow for multiple sets of
+            # scalars. See the ELMo implementation here:
+            # https://github.com/allenai/allennlp/blob/master/allennlp/modules/elmo.py#L115
+            assert len(parse_task_list_arg(args.target_tasks)) <= 1, (
+                "pytorch_transformers_embedding_mode='mix' only supports a single set of "
+                "scalars (but if you need this feature, see the TODO in "
+                "the code!)"
+            )
+            num_layers = self.model.config.num_hidden_layers
+            # Always have one more mixing weight, for lexical layer.
+            self.scalar_mix = scalar_mix.ScalarMix(self.max_layer + 1, do_layer_norm=False)
+
+        self.num_layers = self.model.config.num_hidden_layers
+        if args.pytorch_transformers_max_layer >= 0:
+            self.max_layer = args.pytorch_transformers_max_layer
+            assert self.max_layer <= self.num_layers
+        else:
+            self.max_layer = self.num_layers
+
+    def prepare_output(self, output_seq, lex_seq, hidden_states):
```

Comment:
`output_seq` looks unused here - rename to `unused_output_seq` if it's more convenient to still pass it?

Reply:
Good catch—that got left behind when I merged in your max_layer changes. Removing arg.